### ISSUE TITLE
Questions about the supervised learning.

### ISSUE THREAD

**Issue by CGLemon** [lightvector/KataGo#632]
I try to supervised learning a new 19x19 networks on the current days in my project. The network struct is very similar to Kata Go. The goal is to copy the Kata Go behavior with supervised learning. In this training, I found that different optimizers and parameters may cause the network performance bad. I am confused about that. Hope you may help me. Here are some questions.

1. Adam v.s SGD
From  [one of gcp's comments](https://github.com/gcp/leela-zero/issues/78#issuecomment-353651540). He said that Adam is mush
worse than SGD. It seemd true in my experience. The network policy distribution looks better. But I am still curious about that. Why SGD can beat Adam in this kind of games(Chess-like and  Go)?

2. Learning Rate
In the paper  "Accelerating Self-Play Learning in Go", it picked up a fixed learning rate (=6e-5) and batch size(=256)  except first 5 million samples. Seem we can select a little initial learning rate. But  in my experience, we should select large one. Which initial learning rates are better for supervised learning?

3. Batch Size
According to gcp's comments, he said that small batch had better combine with low learning rate(for example lr= 1e-6, bs=128). But I didn't see any obviously benefit after doing that. In the AlphaZero papers, DeepMind always picked up a large batch size(=2048, 4096). Is gcp right?

**Comment by lightvector** [comment]
1. Adam tries to optimize every parameter taking into account the variance of its variance of gradients. If the variance of gradients on a parameter is unusually low compared to other parameters, Adam will fine-tune the parameter much earlier than SGD will. If the variance is high, Adam will not tune the parameter is as noisy/large-step-size as SGD will. Maybe these behaviors are bad for some reason. I don't know. It could be also that RL vs SL matters too, because in RL, we want the neural net to remain plastic, so that it can re-learn in the future easily when its training distribution shifts.

2. The initial learning rate doesn't matter, it is only very brief (5 million samples is tiny compared to the total data) and the only purpose is to ensure the network doesn't blow up at the very start of training due to large gradients. A gradient clip could also accomplish the same thing.

3. Mostly I've found that batch size doesn't matter too much if you don't use batch norm. For any particular *per sample learning rate*, the training is similar regardless of batch size for typical batch sizes. (Note that if you use per-batch learning rate as normally defined in papers, then you introduce dependence on batch size because the learning rate per sample changes as you change batch size). The reason for using a large batch size is because of hardware - if you are using giant multi-TPUs for your training, a large batch size is mandatory because there is no other way to make the hardware run efficiently. If you don't need a huge batch size on your hardware to make it run efficiently, then don't use one.

**Comment by CGLemon** [comment]
Thanks for your suggestions. But there is something wrong about question 2(Learning Rate) . The initial learning rate what I specify is after first 5 million samples. So the initial learning rate must be big enough? Or it is same result as batch size(variant learning rate will not effect final strength )?

In the theory, if the ```(batch size) / (learning rate)``` is constant, it will be same effect. For example, (bs=128, lr = 0.001) is equal to (bs=256, lr = 0.002). I guess that if the initial learning rate is too low (fix up batch size), may cause the net performance bad. Is it right?

**Comment by lightvector** [comment]
> The initial learning rate what I specify is after first 5 million samples. So the initial learning rate must be big enough? Or it is same result as batch size(variant learning rate will not effect final strength )?
Sorry, I misunderstood you. I recommend you use a different word than "initial" because "initial" means "at the beginning", and so it is easy to interpret you as meaning the learning rate before the first 5 million samples rather than the learning rate after the first 5 million samples, because the first 5 million samples is the beginning of the training.

The general heuristic I've heard is to train as long as you can at close to the highest learning rates you can where the training still is stable and reliable. You tend to reach better optima that way.

> In the theory, if the (batch size) / (learning rate) is constant, it will be same effect. For example, (bs=128, lr = 0.001) is equal to (bs=256, lr = 0.002). 

Yes, if you phrase learning rate as being per-batch instead of per-sample, this is what you would normally expect. The exceptions are when the batch size is so large that even a single step with a given per-sample learning rate is too large of a step to be stable for training, or very early in training when gradients on successive samples are highly correlated, or when batch size itself directly influencing the optimization (e.g. batch norm).
